packages/zoe/zoe.js

```diff
@@ -42,137 +35,238 @@ const makeZoe = async (additionalEndowments = {}) => {
+  // purses. The contract can only do a few things through the Zoe
+  // contract facet. It can propose a reallocation of units,
+  // complete an offer, and can create a new offer itself for
+  // record-keeping and other various purposes.
 
   const makeContractFacet = instanceHandle => {
     const contractFacet = harden({
       /**
        * The contract can propose a reallocation of extents per
-       * player, which will only succeed if the reallocation 1)
+       * offer, which will only succeed if the reallocation 1)
        * conserves rights, and 2) is 'offer-safe' for all parties
        * involved. This reallocation is partial, meaning that it
        * applies only to the extents associated with the offerHandles
-       * that are passed in, rather than applying to all of the
-       * extents in the extentMatrix. We are able to ensure that with
+       * that are passed in. We are able to ensure that with
        * each reallocation, rights are conserved and offer safety is
-       * enforced for all extents, even though the reallocation is
+       * enforced for all offers, even though the reallocation is
        * partial, because once these invariants are true, they will
        * remain true until changes are made.
        * @param  {object[]} offerHandles - an array of offerHandles
-       * @param  {extent[][]} reallocation - a matrix of extents, with
-       * one array of extents per offerHandle. This is likely a subset
-       * of the full extentsMatrix.
+       * @param  {unit[][]} newExtentMatrix - a matrix of extents, with
+       * one array of extents per offerHandle.
        */
-      reallocate: (offerHandles, reallocation) => {
-        const payoutRulesArray = readOnlyState.getPayoutRulesFor(offerHandles);
-        const currentExtents = readOnlyState.getExtentsFor(offerHandles);
-        const extentOpsArray = readOnlyState.getExtentOpsArrayForInstanceHandle(
-          instanceHandle,
-        );
+      reallocate: (offerHandles, newExtentMatrix) => {
+        const { assays } = instanceTable.get(instanceHandle);
+
+        const offers = offerTable.getOffers(offerHandles);
+
+        const payoutRuleMatrix = offers.map(offer => offer.payoutRules);
+        const currentExtentMatrix = offers.map(offer => offer.extents);
+        const extentOpsArray = assayTable.getExtentOpsForAssays(assays);
 
         // 1) ensure that rights are conserved
         insist(
-          areRightsConserved(extentOpsArray, currentExtents, reallocation),
+          areRightsConserved(
+            extentOpsArray,
+            currentExtentMatrix,
+            newExtentMatrix,
+          ),
         )`Rights are not conserved in the proposed reallocation`;
 
         // 2) ensure 'offer safety' for each player
         insist(
-          isOfferSafeForAll(extentOpsArray, payoutRulesArray, reallocation),
+          isOfferSafeForAll(extentOpsArray, payoutRuleMatrix, newExtentMatrix),
```

Comment:
Since the offer safety of each offer is independent, you could decolumnize this further.

Reply:
good point. Let's revisit this when we rework the invariant checks when extents are turned into units.